### PR TITLE
[AIRFLOW-6829][WIP] Auto-apply apply_default

### PR DESCRIPTION
**Pull request by mik-laj**
This decorator is confusing to new users. However, it can be applied automatically without the need to involve users through the use of metaclass.

---
Issue link: [AIRFLOW-6829](https://issues.apache.org/jira/browse/AIRFLOW-6829)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [X] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
Love it

**Comment by potiuk** [comment]
Actualy, I thought a bit and we cannot proceed with this one until we release backporting operators (or we find a way to add BaseMetaclass or apply_defaults automatically via the backported package.

**Review by potiuk — CHANGES_REQUESTED** [review]
It breaks backporting

File: airflow/operators/check_operator.py [hunk 1]
```diff
@@ -123,11 +120,10 @@ class ValueCheckOperator(BaseOperator):
     __mapper_args__ = {
         'polymorphic_identity': 'ValueCheckOperator'
     }
-    template_fields = ('sql', 'pass_value',)  # type: Iterable[str]
+    template_fields = ('sql', 'pass_value',)  # type: Tuple[str, ...]
```
Comment (ashb) [inline airflow/operators/check_operator.py]: Isn't this type the same? (Or this a superset of the previous type)

**Review by ashb — COMMENTED** [review]
Nice!

File: airflow/operators/check_operator.py (hunk 1 above)
Comment (mik-laj) [reply airflow/operators/check_operator.py]: mypy is going crazy in many places now.  It was one of the attempts to fix it.

**Comment by ashb** [comment]
It's horrible, but could we maybe monkey-patch in the metaclass to BashOperator in the back port package? (This is probably a terrible idea even if it's technically possible)

We could do it via a plug-in `load` feature already built in to Airflow (it just gives us any automatically called function>

**Comment by mik-laj** [comment]
We can do it without monkeypatching.  All we have to do is deliver a new BaseOperator class together with operators. We can do it by slightly modifying the source code.

**Comment by potiuk** [comment]
I think we will eventually have to slightly modify the source code of the operators when preparing backport packages indeed and In this case I agree monkey patching is worse idea :). 

The idea is to have class deriving from the BaseOperator and implement MetaClass and have all the backported operators to derive from this class instead. We can do it by modifying the source code (slightly hacky) or by modifying the Concrete Syntax Tree via https://pybowler.io/

**Comment by mik-laj** [comment]
Let's wait with this PR until we stop relasing the package of backported operators. This PR shows that it is possible to use the metaclass to delete apply_default decorator. When we give up support for packages with operators, I will continue working on this PR.

**Comment by turbaszek** [comment]
@mik-laj I think it would be awesome to have this in Airflow 2.0. @ashb @potiuk @kaxil WDYT?

**Comment by ashb** [comment]
Yeah I agree - missing this out was a common mistske!

**Review by turbaszek — APPROVED** [review]
Look good to me! The change in automl is similar to this one #10869

**Comment by mik-laj** [comment]
I finished this PR, but there is a minor problem if we want to be compatible with Airflow 1.10 and still release packages for Airflow. 1.10. 

The `airflow.utils.decorators.apply_defaults` decorator displays deprecation warnings (it's okey).  The problem is that we should update all operators and delete invocation of this decorator so that these warnings will not appear. But we cannot do that if we want to keep release backport packages, because in Airflow 1.10, this decorator is still required. We also can't add this decorator with Bowler, because Bowler could get lost in the class hierarchy and it's too complex so I can't trust him.

I can see several solutions:
- Merge unchanged and messages will be displayed in Airflow 2.0 alpha and Breeze. 
- Add a bowler refactor for the provider packages for Airflow 2.0 that will delete this decorator
- Conditionally display warnings for Breeze. Once the user is in the breeze then they will not see the warning.

@ashb @kaxil @potiuk WDYT? It would be nice if we could include these changes into the next alpha.

**Comment by potiuk** [comment]
My current preference is those two together:

* Add a bowler refactor for the provider packages for Airflow 2.0 that will delete this decorator. Previously, provider package didn't have any bowler refactors, so that would be new.
* Conditionally display warnings for Breeze. Once the user is in the breeze then they will not see the warning. 

This is very simple to do, it will connect both  - no "apply_defaults" in "production" code and easy development in Breeze. For local development (we could even leave with displaying the warnings locally with "those apply_defaults" are here now because we need them until we release backports but we intend to remove it once backport-providers are gone (which should happen in Feb/March according to our plans). We can talk about today at the 2.0 meeting/

**Comment by ashb** [comment]
> Conditionally display warnings for Breeze

What about all the users who don't use breeze?

**Comment by dimberman** [comment]
@mik-laj is this PR still relevant or can we close it?

**Comment by mik-laj** [comment]
It is still relevant.

**Comment by mik-laj** [comment]
superseded by  https://github.com/apache/airflow/pull/15044

**Pull request closed without merging**